### PR TITLE
[FLINK-13433][table-planner-blink]  Do not fetch data from LookupableTableSource if the JoinKey in left side of LookupJoin contains null value.

### PR DESCRIPTION
**Pull request by beyond1920**
## What is the purpose of the change

For LookupJoin, if joinKey in left side of a LeftOuterJoin/InnerJoin contains null values, there is no need to fetch data from `LookupableTableSource`. 
However, we don't shortcut the fetch function under the case at present, the correctness of results depends on the `TableFunction` implementation of each `LookupableTableSource`.

## Brief change log

  - correct the behavior of `LookupJoin` where lookupKeys contains null.
  - Add more testcases.

## Verifying this change
testcases

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature?  no
  - If yes, how is the feature documented? not applicable

### CODE REVIEW

**Comment by beyond1920** [comment]
cc @wuchong @lincoln-lil

**Comment by JingsongLi** [comment]
I am interested to sql: `SELECT T.id, T.len, T.content, D.name FROM T JOIN userTable for system_time as of T.proctime AS D ON T.id = D.id OR (T.id is null and D.id is null)`.
Does it work?

**Comment by beyond1920** [comment]
@JingsongLi At first glance, I guess the query is not supported. 
However, after run it actually, The query could be executed, but results in error outputs……
Good catch, I will looks into the problem.

File: flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/codegen/LookupJoinCodeGenerator.scala [hunk 1]
```diff
@@ -133,8 +137,13 @@ object LookupJoinCodeGenerator {
     val body =
       s"""
          |$prepareCode
-         |$DELEGATE delegates = new $DELEGATE($DEFAULT_COLLECTOR_TERM);
-         |$lookupFunctionTerm.eval(delegates.getCompletableFuture(), $parameters);
+         |if ($parametersCheckNullCode) {
+         |  $DEFAULT_COLLECTOR_TERM.complete(new java.util.ArrayList());
```
Comment (lincoln-lil) [inline flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/codegen/LookupJoinCodeGenerator.scala]: use `Collections.emptyList()` is better

File: flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/plan/rules/physical/common/CommonLookupJoinRule.scala [hunk 2]
```diff
@@ -86,6 +91,21 @@ trait CommonLookupJoinRule {
...
+
+    val filterNulls: Array[Boolean] = {
+      val filterNulls = new util.ArrayList[java.lang.Boolean]
+      JoinUtil.createJoinInfo(join.getLeft, join.getRight, join.getCondition, filterNulls)
+      filterNulls.map(_.booleanValue()).toArray
+    }
+
+    if (filterNulls.contains(false)) {
+      throw new TableException(
+        "IS NOT DISTINCT FROM (or an expanded version) is not supported in LookupJoin.")
```
Comment (lincoln-lil) [inline flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/plan/rules/physical/common/CommonLookupJoinRule.scala]: Give the detail condition maybe more clearly to users?

File: flink-table/flink-table-planner-blink/src/test/scala/org/apache/flink/table/planner/plan/batch/sql/join/LookupJoinTest.scala [hunk 3]
```diff
@@ -85,6 +85,26 @@ class LookupJoinTest extends TableTestBase {
     )
   }
 
+  @Test
+  def testNotDistinctFromInJoinCondition(): Unit = {
+
+    // does not support join condition contains `IS NOT  DISTINCT`
```
Comment (lincoln-lil) [inline flink-table/flink-table-planner-blink/src/test/scala/org/apache/flink/table/planner/plan/batch/sql/join/LookupJoinTest.scala]: `NOT  DISTINCT`  rm one space.

**Comment by beyond1920** [comment]
@wuchong 
> could you also update InMemoryAsyncLookupFunction#eval and InMemoryLookupFunction#eval to throw exception when one of the argument is null？
Hi, JarkWu, InMemoryAsyncLookupFunction and InMemoryLookupFunction lookup records with null value on the lookup key field when argument contains null, which complies with the contract mentioned in https://github.com/apache/flink/pull/9335/files, I think the behavior is also reasonable.

**Comment by wuchong** [comment]
I agree with you @beyond1920 , however, I want to have some test to cover we didn't push any nulls into `LookupFunction`. That's why I want to add such check in `InMemoryLookupFunction`. We can remove the check when we support push `is null` into LookupFunction?

**Comment by beyond1920** [comment]
@wuchong , Ok

**Comment by wuchong** [comment]
Thanks @beyond1920 , looks good to me now.

**Comment by KurtYoung** [comment]
BTW, could you also open a jira for empty key for hbase lookup function? hbase forbidden empty row key.

**Comment by beyond1920** [comment]
@KurtYoung The hbase lookup function is updated in https://github.com/apache/flink/pull/9335.

**Pull request closed without merging**